File: ios/YamabikoChat/Data/Repositories/ModelsDevCatalogRepository.swift
import Combine
import Foundation

private actor ModelsDevFetchCoordinator {
    private var task: Task<CatalogLoadState, Never>?

    func resolve(forceRefresh: Bool, operation: @Sendable @escaping () async -> CatalogLoadState) async -> CatalogLoadState {
        if forceRefresh { task?.cancel(); task = nil }
        if let task { return await task.value }
        let next = Task { await operation() }
        task = next
        let value = await next.value
        task = nil
        return value
    }
}

final class ModelsDevCatalogRepository: @unchecked Sendable {
    private static let catalogURL = URL(string: "https://models.dev/catalog.json")!
    private static let cacheTTL: TimeInterval = 24 * 60 * 60
    private static let etagKey = "models_dev_catalog_etag"
    private static let fetchedAtKey = "models_dev_catalog_fetched_at"

    private let session: URLSession
    private let defaults: UserDefaults
    private let cacheURL: URL
    private let stateSubject: CurrentValueSubject<CatalogLoadState, Never>
    private let coordinator = ModelsDevFetchCoordinator()

    init(session: URLSession = .shared, defaults: UserDefaults = .standard, cacheURL: URL? = nil) {
        self.session = session
        self.defaults = defaults
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        self.cacheURL = cacheURL ?? base.appendingPathComponent("models_dev_catalog_cache.json")
        let providers = Self.readCache(at: self.cacheURL)
        let date = defaults.object(forKey: Self.fetchedAtKey) as? Date
        stateSubject = CurrentValueSubject(CatalogLoadState(
            availability: providers.isEmpty ? .idle : .stale,
            providers: providers,
            lastUpdated: date,
            error: nil
        ))
    }

    var statePublisher: AnyPublisher<CatalogLoadState, Never> { stateSubject.eraseToAnyPublisher() }
    func currentState() -> CatalogLoadState { stateSubject.value }

    func provider(for reference: ProviderReference) -> CatalogProvider? {
        guard let id = reference.modelsDevID else { return nil }
        return stateSubject.value.providers.first { $0.id == id }
    }

    func load(forceRefresh: Bool = false) async -> CatalogLoadState {
        let current = stateSubject.value
        if !forceRefresh,
           !current.providers.isEmpty,
           let updated = current.lastUpdated,
           Date().timeIntervalSince(updated) < Self.cacheTTL {
            let ready = CatalogLoadState(availability: .ready, providers: current.providers, lastUpdated: updated)
            stateSubject.send(ready)
            return ready
        }
        stateSubject.send(CatalogLoadState(availability: .loading, providers: current.providers, lastUpdated: current.lastUpdated))
        let result = await coordinator.resolve(forceRefresh: forceRefresh) { [self] in
            await fetch(previous: current)
        }
        stateSubject.send(result)
        return result
    }

    private func fetch(previous: CatalogLoadState) async -> CatalogLoadState {
        do {
            var request = URLRequest(url: Self.catalogURL, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 30)
            if let etag = defaults.string(forKey: Self.etagKey) { request.setValue(etag, forHTTPHeaderField: "If-None-Match") }
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
            if http.statusCode == 304, !previous.providers.isEmpty {
                let date = Date()
                defaults.set(date, forKey: Self.fetchedAtKey)
                return CatalogLoadState(availability: .ready, providers: previous.providers, lastUpdated: date)
            }
            guard (200 ... 299).contains(http.statusCode) else {
                throw ProviderClientError.httpStatus(http.statusCode, String(data: data, encoding: .utf8) ?? "")
            }
            let providers = try Self.parseCatalog(data)
            guard !providers.isEmpty else { throw ProviderClientError.parseFailure("models.dev returned no usable providers") }
            try FileManager.default.createDirectory(at: cacheURL.deletingLastPathComponent(), withIntermediateDirectories: true)
            try JSONEncoder().encode(providers).write(to: cacheURL, options: .atomic)
            let date = Date()
            defaults.set(date, forKey: Self.fetchedAtKey)
            defaults.set(http.value(forHTTPHeaderField: "ETag"), forKey: Self.etagKey)
            DiagnosticsLogger.log("models.dev catalog updated providers=\(providers.count)", category: .network)
            return CatalogLoadState(availability: .ready, providers: providers, lastUpdated: date)
        } catch {
            DiagnosticsLogger.log("models.dev catalog update failed", category: .network, error: error)
            return CatalogLoadState(
                availability: previous.providers.isEmpty ? .error : .stale,
                providers: previous.providers,
                lastUpdated: previous.lastUpdated,
                error: error.localizedDescription
            )
        }
    }

    static func parseCatalog(_ data: Data) throws -> [CatalogProvider] {
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let rawProviders = root["providers"] as? [String: Any]
        else { throw ProviderClientError.parseFailure("catalog.json is missing providers") }

        return rawProviders.compactMap { providerID, rawValue -> CatalogProvider? in
            guard providerID.lowercased() != "openrouter",
                  let value = rawValue as? [String: Any],
                  let name = value["name"] as? String,
                  let npm = value["npm"] as? String,
                  let rawModels = value["models"] as? [String: Any]
            else { return nil }
            let models = rawModels.compactMap { modelID, rawModel -> CatalogModel? in
                guard let model = rawModel as? [String: Any],
                      (model["status"] as? String)?.lowercased() != "deprecated"
                else { return nil }
                let modalities = model["modalities"] as? [String: Any]
                let outputs = modalities?["output"] as? [String] ?? []
                guard outputs.contains(where: { $0.caseInsensitiveCompare("text") == .orderedSame }) else { return nil }
                let limits = model["limit"] as? [String: Any]
                let cost = model["cost"] as? [String: Any]
                let modelProvider = model["provider"] as? [String: Any]
                let reasoningOptions = (model["reasoning_options"] as? [[String: Any]] ?? []).compactMap { option in
                    (option["type"] as? String).map { type in
                        let values = (option["values"] as? [Any] ?? []).compactMap { $0 as? String }
                        return CatalogReasoningOption(type: type, values: values)
                    }
                }
                return CatalogModel(
                    id: modelID,
                    name: model["name"] as? String ?? modelID,
                    description: model["description"] as? String,
                    family: model["family"] as? String,
                    attachment: model["attachment"] as? Bool,
                    reasoning: model["reasoning"] as? Bool,
                    reasoningOptions: reasoningOptions,
                    toolCall: model["tool_call"] as? Bool,
                    structuredOutput: model["structured_output"] as? Bool,
                    temperature: model["temperature"] as? Bool,
                    inputModalities: modalities?["input"] as? [String] ?? [],
                    outputModalities: outputs,
                    releaseDate: model["release_date"] as? String,
                    lastUpdated: model["last_updated"] as? String,
                    limits: CatalogLimits(context: number(limits?["context"]), input: number(limits?["input"]), output: number(limits?["output"])),
                    cost: CatalogCost(
                        inputPerMillion: decimal(cost?["input"]), outputPerMillion: decimal(cost?["output"]),
                        reasoningPerMillion: decimal(cost?["reasoning"]), cacheReadPerMillion: decimal(cost?["cache_read"]),
                        cacheWritePerMillion: decimal(cost?["cache_write"])
                    ),
                    providerContract: CatalogModelProviderContract(
                        npm: modelProvider?["npm"] as? String ?? npm,
                        api: modelProvider?["api"] as? String ?? value["api"] as? String,
                        shape: modelProvider?["shape"] as? String,
                        provenance: modelProvider == nil ? "provider" : "model"
                    )
                )
            }.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
            guard !models.isEmpty else { return nil }
            return CatalogProvider(
                id: providerID.lowercased(), name: name, npm: npm, api: value["api"] as? String,
                env: value["env"] as? [String] ?? [], documentationURL: value["doc"] as? String, models: models
            )
        }.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    private static func readCache(at url: URL) -> [CatalogProvider] {
        (try? Data(contentsOf: url)).flatMap { try? JSONDecoder().decode([CatalogProvider].self, from: $0) } ?? []
    }

    private static func number(_ value: Any?) -> Int? { (value as? NSNumber)?.intValue }
    private static func decimal(_ value: Any?) -> Double? { (value as? NSNumber)?.doubleValue }
}

File: app/src/main/java/com/porarri/yamabikochat/data/modelsdev/ModelsDevCatalogRepository.kt
package com.porarri.yamabikochat.data.modelsdev

import android.content.Context
import android.util.AtomicFile
import com.porarri.yamabikochat.utils.DiagnosticsLogger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.util.concurrent.TimeUnit

class ModelsDevCatalogRepository(
    context: Context,
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build(),
    private val now: () -> Long = System::currentTimeMillis
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val cacheFile = File(context.filesDir, "models_dev_catalog_cache.json")
    private val metadataPrefs = context.getSharedPreferences("models_dev_catalog_metadata", Context.MODE_PRIVATE)
    private val fetchMutex = Mutex()
    private val _state = MutableStateFlow(CatalogLoadState())
    val state: StateFlow<CatalogLoadState> = _state.asStateFlow()

    suspend fun load(forceRefresh: Boolean = false): CatalogLoadState = fetchMutex.withLock {
        val cached = readCache()
        val fetchedAt = metadataPrefs.getLong(KEY_FETCHED_AT, 0L).takeIf { it > 0 }
        val cacheFresh = cached.isNotEmpty() && fetchedAt != null && now() - fetchedAt < CACHE_TTL_MS
        if (!forceRefresh && cacheFresh) {
            return@withLock CatalogLoadState(CatalogAvailability.READY, cached, fetchedAt).also { _state.value = it }
        }

        _state.value = CatalogLoadState(CatalogAvailability.LOADING, cached, fetchedAt)
        val result = withContext(Dispatchers.IO) { fetchCatalog(cached, fetchedAt) }
        _state.value = result
        result
    }

    fun providers(): List<CatalogProvider> = state.value.providers

    fun provider(reference: ProviderReference): CatalogProvider? =
        reference.modelsDevId?.let { id -> providers().firstOrNull { it.id == id } }

    suspend fun providerOrLoad(reference: ProviderReference): CatalogProvider? {
        if (!reference.isModelsDev) return null
        provider(reference)?.let { return it }
        return load().providers.firstOrNull { it.id == reference.modelsDevId }
    }

    private fun fetchCatalog(cached: List<CatalogProvider>, fetchedAt: Long?): CatalogLoadState {
        return try {
            val requestBuilder = Request.Builder().url(CATALOG_URL).get()
            metadataPrefs.getString(KEY_ETAG, null)?.let { requestBuilder.header("If-None-Match", it) }
            client.newCall(requestBuilder.build()).execute().use { response ->
                if (response.code == 304 && cached.isNotEmpty()) {
                    val updatedAt = now()
                    metadataPrefs.edit().putLong(KEY_FETCHED_AT, updatedAt).apply()
                    return CatalogLoadState(CatalogAvailability.READY, cached, updatedAt)
                }
                if (!response.isSuccessful) error("models.dev returned HTTP ${response.code}")
                val body = response.body?.string().orEmpty()
                val parsed = parseCatalog(body)
                require(parsed.isNotEmpty()) { "models.dev catalog contains no usable providers" }
                writeCache(parsed)
                val updatedAt = now()
                metadataPrefs.edit()
                    .putLong(KEY_FETCHED_AT, updatedAt)
                    .putString(KEY_ETAG, response.header("ETag"))
                    .apply()
                DiagnosticsLogger.log("models.dev catalog updated providers=${parsed.size}")
                CatalogLoadState(CatalogAvailability.READY, parsed, updatedAt)
            }
        } catch (error: Exception) {
            DiagnosticsLogger.log("models.dev catalog update failed", error)
            CatalogLoadState(
                availability = if (cached.isEmpty()) CatalogAvailability.ERROR else CatalogAvailability.STALE,
                providers = cached,
                lastUpdatedEpochMs = fetchedAt,
                error = error.message ?: "models.dev catalog update failed"
            )
        }
    }

    internal fun parseCatalog(raw: String): List<CatalogProvider> {
        val root = json.parseToJsonElement(raw).jsonObject
        val providersObject = (root["providers"] ?: error("catalog.json is missing providers")).jsonObject
        return providersObject.entries.asSequence()
            .filter { (id, _) -> id.lowercase() != "openrouter" }
            .mapNotNull { (providerId, element) -> parseProvider(providerId, element.jsonObject) }
            .sortedBy { it.name.lowercase() }
            .toList()
    }

    private fun parseProvider(id: String, value: JsonObject): CatalogProvider? {
        val name = value.string("name") ?: return null
        val npm = value.string("npm") ?: return null
        val api = value.string("api")
        val models = value["models"]?.jsonObject?.entries.orEmpty().mapNotNull { (modelId, modelValue) ->
            parseModel(modelId, modelValue.jsonObject, npm, api)
        }.sortedBy { it.name.lowercase() }
        if (models.isEmpty()) return null
        return CatalogProvider(
            id = id.lowercase(),
            name = name,
            npm = npm,
            api = api,
            env = value.stringList("env"),
            documentationUrl = value.string("doc"),
            models = models
        )
    }

    private fun parseModel(id: String, value: JsonObject, providerNpm: String, providerApi: String?): CatalogModel? {
        if (value.string("status")?.lowercase() == "deprecated") return null
        val outputs = value["modalities"]?.jsonObject?.stringList("output").orEmpty()
        if (outputs.none { it.equals("text", ignoreCase = true) }) return null
        val reasoningOptions = value["reasoning_options"]?.jsonArray?.mapNotNull { option ->
            val objectValue = option.jsonObject
            objectValue.string("type")?.let { CatalogReasoningOption(it, objectValue.stringList("values")) }
        }.orEmpty()
        val limit = value["limit"]?.jsonObject
        val cost = value["cost"]?.jsonObject
        val modelProvider = value["provider"]?.jsonObject
        return CatalogModel(
            id = id,
            name = value.string("name") ?: id,
            description = value.string("description"),
            family = value.string("family"),
            attachment = value.optionalBool("attachment"),
            reasoning = value.optionalBool("reasoning"),
            reasoningOptions = reasoningOptions,
            toolCall = value.optionalBool("tool_call"),
            structuredOutput = value.optionalBool("structured_output"),
            temperature = value.optionalBool("temperature"),
            inputModalities = value["modalities"]?.jsonObject?.stringList("input").orEmpty(),
            outputModalities = outputs,
            releaseDate = value.string("release_date"),
            lastUpdated = value.string("last_updated"),
            limits = CatalogLimits(limit?.long("context"), limit?.long("input"), limit?.long("output")),
            cost = CatalogCost(
                cost?.double("input"), cost?.double("output"), cost?.double("reasoning"),
                cost?.double("cache_read"), cost?.double("cache_write")
            ),
            providerContract = CatalogModelProviderContract(
                npm = modelProvider?.string("npm") ?: providerNpm,
                api = modelProvider?.string("api") ?: providerApi,
                shape = modelProvider?.string("shape"),
                provenance = if (modelProvider == null) "provider" else "model"
            )
        )
    }

    private fun readCache(): List<CatalogProvider> = runCatching {
        if (!cacheFile.exists()) emptyList() else json.decodeFromString<List<CatalogProvider>>(cacheFile.readText())
    }.onFailure { DiagnosticsLogger.log("models.dev cache read failed", it) }.getOrDefault(emptyList())

    private fun writeCache(providers: List<CatalogProvider>) {
        val atomicFile = AtomicFile(cacheFile)
        val stream = atomicFile.startWrite()
        try {
            stream.write(json.encodeToString(providers).toByteArray(Charsets.UTF_8))
            atomicFile.finishWrite(stream)
        } catch (error: Exception) {
            atomicFile.failWrite(stream)
            throw error
        }
    }

    private fun JsonObject.string(key: String): String? = this[key]?.jsonPrimitive?.contentOrNull
    private fun JsonObject.optionalBool(key: String): Boolean? = this[key]?.jsonPrimitive?.booleanOrNull
    private fun JsonObject.long(key: String): Long? = this[key]?.jsonPrimitive?.longOrNull
    private fun JsonObject.double(key: String): Double? = this[key]?.jsonPrimitive?.doubleOrNull
    private fun JsonObject.stringList(key: String): List<String> =
        this[key]?.jsonArray?.mapNotNull { it.jsonPrimitive.contentOrNull }.orEmpty()

    private companion object {
        const val CATALOG_URL = "https://models.dev/catalog.json"
        const val KEY_FETCHED_AT = "fetched_at"
        const val KEY_ETAG = "etag"
        const val CACHE_TTL_MS = 24 * 60 * 60 * 1000L
    }
}
